### PR TITLE
Improve CborWriter buffer growth

### LINKED ISSUE #92425: CborWriter should use doubling for buffer growth
Currently, `CborWriter` increases its buffer 1024 bytes at a time:

https://github.com/dotnet/runtime/blob/51498964d35068ded252ec98baf4fe6ebf6a4612/src/libraries/System.Formats.Cbor/src/System/Formats/Cbor/Writer/CborWriter.cs#L210-L212

For large CBOR documents, this can be result in a lot of array resizing. We should instead change the strategy to do doubling.

### PR DESCRIPTION
**Pull request by MichalPetryka**
Fixes #92425.

### CODE REVIEW

File: src/libraries/System.Formats.Cbor/src/System/Formats/Cbor/Writer/CborWriter.cs
Comment (vcsjones) [inline src/libraries/System.Formats.Cbor/src/System/Formats/Cbor/Writer/CborWriter.cs:214]: This package builds for netstandard2.0. I don't believe `MaxLength` is available there.

Comment (hamarb123) [inline src/libraries/System.Formats.Cbor/src/System/Formats/Cbor/Writer/CborWriter.cs:214]: would it be possible for it to get here when trying to expand to larger than `Array.MaxLength`, shouldn't there be some condition to throw here? e.g., you could just check if `requiredCapacity > Array.MaxLength` and throw if it is somewhere.

File: src/libraries/System.Formats.Cbor/src/System/Formats/Cbor/Writer/CborWriter.cs
Comment (hamarb123) [inline src/libraries/System.Formats.Cbor/src/System/Formats/Cbor/Writer/CborWriter.cs:218]: would it be worth using `GC.AllocateUninitializedArray` here?

File: src/libraries/System.Formats.Cbor/src/System/Formats/Cbor/Writer/CborWriter.cs
Comment (MichalPetryka) [reply src/libraries/System.Formats.Cbor/src/System/Formats/Cbor/Writer/CborWriter.cs:214]: AFAIR `new byte[]` will throw on such length.

Comment (hamarb123) [reply src/libraries/System.Formats.Cbor/src/System/Formats/Cbor/Writer/CborWriter.cs:214]: ahh, I missed the next line lol - please close since I can't for some reason

File: src/libraries/System.Formats.Cbor/src/System/Formats/Cbor/Writer/CborWriter.cs
Comment (MichalPetryka) [reply src/libraries/System.Formats.Cbor/src/System/Formats/Cbor/Writer/CborWriter.cs:218]: I'll let the maintainers decide whether it makes sense here.

File: src/libraries/System.Formats.Cbor/src/System/Formats/Cbor/Writer/CborWriter.cs
Comment (MichalPetryka) [reply src/libraries/System.Formats.Cbor/src/System/Formats/Cbor/Writer/CborWriter.cs]: Ah right, should I replace it with some hardcoded constant like 1GB then?

Comment (hamarb123) [reply src/libraries/System.Formats.Cbor/src/System/Formats/Cbor/Writer/CborWriter.cs]: I would replace it with the `Array.MaxValue` constant where it's unavailable; I think that's the correct value for basically all the runtimes we've had anyway isn't it?

File: src/libraries/System.Formats.Cbor/src/System/Formats/Cbor/Writer/CborWriter.cs
Comment (bartonjs) [inline src/libraries/System.Formats.Cbor/src/System/Formats/Cbor/Writer/CborWriter.cs:214]: Braces are always appreciated :)

```suggestion
                {
                    newCapacity = requiredCapacity;
                }
```

File: src/libraries/System.Formats.Cbor/src/System/Formats/Cbor/Writer/CborWriter.cs
Comment (bartonjs) [reply src/libraries/System.Formats.Cbor/src/System/Formats/Cbor/Writer/CborWriter.cs:218]: AllocateUninitializedArray would probably be OK, except it's not available for .NET Standard 2.0.  I don't think it's important enough here that it's worth a `#if`, so just always using `new byte[newCapacity]` seems fine.

File: src/libraries/System.Formats.Cbor/src/System/Formats/Cbor/Writer/CborWriter.cs
Comment (vcsjones) [reply src/libraries/System.Formats.Cbor/src/System/Formats/Cbor/Writer/CborWriter.cs]: For both .NET and .NET Framework, the array length limit when the array is a byte[] is 0x7FFFFFC7. It's probably reasonable to do 

```C#
const int MaxArrayLength = 0x7FFFFFC7; // Array.MaxLength
```

Or if we don't feel like burning it in as a const for the latest runtime:


```C#
#if NET
    int maxArrayLength = Array.Length;
#else
    int MaxArrayLength = 0x7FFFFFC7; // Array.MaxLength
#endif
```

File: src/libraries/System.Formats.Cbor/src/System/Formats/Cbor/Writer/CborWriter.cs
Comment (eiriktsarpalis) [inline src/libraries/System.Formats.Cbor/src/System/Formats/Cbor/Writer/CborWriter.cs:213]: Maybe add an assertion validating that we're picking the right constant.
```suggestion
                const uint MaxArrayLength = 0x7FFFFFC7; // Array.MaxLength
#if NETCOREAPP
                Debug.Assert(MaxArrayLength == Array.MaxLength);
#endif
```

File: src/libraries/System.Formats.Cbor/src/System/Formats/Cbor/Writer/CborWriter.cs
Comment (eiriktsarpalis) [inline src/libraries/System.Formats.Cbor/src/System/Formats/Cbor/Writer/CborWriter.cs:214]: This check seems to be redundant, either the doubling will exceed `MaxArrayLength` or it will overflow. The current initial capacity + doubling algorithm seems to suggest that the latter is going to happen.

File: src/libraries/System.Formats.Cbor/src/System/Formats/Cbor/Writer/CborWriter.cs
Comment (eiriktsarpalis) [reply src/libraries/System.Formats.Cbor/src/System/Formats/Cbor/Writer/CborWriter.cs:219]: Why not just use Array.Resize as before?

File: src/libraries/System.Formats.Cbor/src/System/Formats/Cbor/Writer/CborWriter.cs
Comment (eiriktsarpalis) [inline src/libraries/System.Formats.Cbor/src/System/Formats/Cbor/Writer/CborWriter.cs:205]: Apropos, I wrote a lot of this code and it hindsight it seems like I didn't account for not inlining throws in hot-path methods. We should consider refactoring into throw helpers at some point, though clearly not in scope for this PR.

File: src/libraries/System.Formats.Cbor/src/System/Formats/Cbor/Writer/CborWriter.cs
Comment (MichalPetryka) [reply src/libraries/System.Formats.Cbor/src/System/Formats/Cbor/Writer/CborWriter.cs:222]: > Presumably the original buffer is close to full so we're not saving a lot by only copying the written slice.

The intent was to avoid copying empty space here, since copying is really the biggest cost on resizing here.

**Review by eiriktsarpalis — APPROVED** [review]
Thanks

**Comment by vcsjones** [comment]
@MichalPetryka Can you please fix the merge conflict?

**Comment by MichalPetryka** [comment]
> Can you please fix the merge conflict?

Done

**Pull request merged**
